scylla-cql/src/types/deserialize/result.rs

```diff
@@ -136,46 +140,221 @@ where
-
-        let mut row1 = iter.next().unwrap().unwrap();
-        let c11 = row1.next().unwrap().unwrap();
-        assert_eq!(c11.slice.unwrap().as_slice(), CELL1);
-        let c12 = row1.next().unwrap().unwrap();
-        assert_eq!(c12.slice.unwrap().as_slice(), CELL2);
-        assert!(row1.next().is_none());
-
-        let mut row2 = iter.next().unwrap().unwrap();
-        let c21 = row2.next().unwrap().unwrap();
-        assert_eq!(c21.slice.unwrap().as_slice(), CELL2);
-        let c22 = row2.next().unwrap().unwrap();
-        assert_eq!(c22.slice.unwrap().as_slice(), CELL1);
-        assert!(row2.next().is_none());
+    impl<'frame, 'metadata> LendingIterator for RowIterator<'frame, 'metadata> {
+        type Item<'borrow> = ColumnIterator<'borrow, 'borrow>
+        where
+            Self: 'borrow;
 
-        assert!(iter.next().is_none());
+        fn lend_next(&mut self) -> Option<Result<ColumnIterator<'_, '_>, DeserializationError>> {
+            self.next()
+        }
+    }
+
+    impl LendingIterator for RawRowsLendingIterator {
+        type Item<'borrow> = ColumnIterator<'borrow, 'borrow>;
+
+        fn lend_next(&mut self) -> Option<Result<ColumnIterator<'_, '_>, DeserializationError>> {
+            self.next()
+        }
     }
 
     #[test]
-    fn test_row_iterator_too_few_rows() {
-        let raw_data = serialize_cells([Some(CELL1), Some(CELL2)]);
-        let specs = [spec("b1", ColumnType::Blob), spec("b2", ColumnType::Blob)];
-        let mut iter = RowIterator::new(2, &specs, FrameSlice::new(&raw_data));
+    fn test_row_iterators_basic_parse() {
+        static RAW_DATA: LazyLock<Bytes> =
+            LazyLock::new(|| serialize_cells([Some(CELL1), Some(CELL2), Some(CELL2), Some(CELL1)]));
+
+        let raw_data = RAW_DATA.deref();
+
+        let specs = vec![spec("b1", ColumnType::Blob), spec("b2", ColumnType::Blob)];
+
+        let row_iter = RowIterator::new(
+            2,
+            // Those leaks are required because of a compiler bug-limitation about GATs:
+            // https://blog.rust-lang.org/2022/10/28/gats-stabilization.html#implied-static-requirement-from-higher-ranked-trait-bounds
+            // the following type higher-ranked lifetime constraint implies 'static lifetime.
+            //
+            // I: for<'item> LendingIterator<Item<'item> = ColumnIterator<'item>>,
+            //
+            // The bug is said to be a lot of effort to fix, so in tests let's just use `leak`
+            // to obtain 'static references.
+            Vec::leak(specs.clone()),
+            FrameSlice::new(Box::leak(Box::new(raw_data.clone()))),
+        );
+        let lending_row_iter =
+            RawRowsLendingIterator::new(DeserializedMetadataAndRawRows::new_for_test(
+                ResultMetadata::new_for_test(specs.len(), specs),
+                2,
+                raw_data.clone(),
+            ));
+        check(row_iter);
+        check(lending_row_iter);
```

Comment:
What do those tests do that normal code won't?
If there is a need to leak in normal code then it is not acceptable.

Reply:
> What do those tests do that normal code won't?

These tests use a lending iterator trait. GATs currently have a compiler limitation that 'static lifetime is required.

Reply:
Will normal code not use lending iterators? If so, why do we need them and test them?

Reply:
I use them for convenience in tests, not to duplicate code for both iterators.
Of course, I could work around this. But isn't it nice to have GATs in our codebase? :sparkles:

Reply:
Its not if it requires leaking.

Reply:
I replaced leaks with `lazy_static`.